### PR TITLE
Add domain settings for restricting UCR expressions

### PR DESCRIPTION
**Pull request by AmitPhulera**
## Product Description

The PR implements the first part of the [technical spec](https://docs.google.com/document/d/118z4wiClwgH-r_-6a77WBf0E3uLPaX-lyL-iCX1VoD0/edit) 
The current PR
- Adds an attribute in domain settings
- Adds UI to update it
- A management command to populate the settings on domains with UCR feature flag turned on.
- Rename user-facing text of UCR as per [this](https://docs.google.com/document/d/19oscr4oHnaZiRE7IG6oyBsAR0kKLQViF4Y-ufXggjAg/edit) doc
Screenshot 2022-02-14 at 11 23 21 AM

Review by commit 🐟 
## Technical Summary

https://dimagi-dev.atlassian.net/browse/SAAS-13142
### Rollback instructions

- [x] This PR can be reverted after deploy with no further considerations

### Labels & Review
- [x] Risk label is set correctly
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change

### CODE REVIEW

**Review by snopoke — COMMENTED** [review]
Couple of comments:
* The domain model is already very bloated. Can we not make this a separate model?
* Instead updating all domains to set the default value why not just use the default value unless a domain has set a custom value i.e. use the config value defined if it exists otherwise use the default?

**Comment by AmitPhulera** [comment]
Thanks, @snopoke for sharing the suggestions.

> The domain model is already very bloated. Can we not make this a separate model?

We surely can. When I think about it, two approaches come to my mind-
- Keep this model separated altogether, it would not deal with Domain object in any way.
- Create a new model and add reference of the newly created model in the Domain object.

I personally think the first approach tackles the bloating problem that you have mentioned it well but it would isolate this permission completely


> Instead updating all domains to set the default value why not just use the default value unless a domain has set a custom value i.e. use the config value defined if it exists otherwise use the default?

This is a great suggestion. I would implement it after we finalize the approach with which we want to go.

**Comment by snopoke** [comment]
> Keep this model separated altogether, it would not deal with Domain object in any way.

:+1:

**Comment by AmitPhulera** [comment]
Hey @snopoke, 
I have made the required changes

File: corehq/apps/domain/utils.py [hunk 1]
```diff
@@ -124,3 +124,21 @@ def get_serializable_wire_invoice_general_credit(general_credit):
         }]
 
     return []
+
+
+def get_dte_expressions(domain_name):
```
[the remark this answers is not in the document]
Comment (AmitPhulera) [reply corehq/apps/domain/utils.py:129]: To ensure that things work consistently with Solutions, we are allowing all the expressions by default. We would be restricting certain expressions for SaaS partners manually. So that is why it is returning a full list of all restricted expressions by default.

> I would think it should return an empty list if the model doesn't exist. In this case, I'd also recommend renaming it to 'get_allowed_ucr_expressions'

I will update the name.

File: corehq/apps/domain/models.py [hunk 2]
```diff
@@ -1047,6 +1059,23 @@ def update_calculations(cls, domain, property_to_update):
         obj.save(update_fields=[property_to_update])
 
 
+class AllowedDTEExpressionSettings(models.Model):
+    """
+    Model contains DTE(aka UCR) expressions settings for a domain.
```
[the remark this answers is not in the document]
Comment (AmitPhulera) [reply corehq/apps/domain/models.py:1064]: In the beginning, I had a similar question in my mind, whether to go with UCR or with DTE. The reasoning behind going with this name is that we are planning to rename UCR with Data Transformation Engine. The cosmetic changes have started with this PR. 
I was not sure if we would rename it in code in near future but that seemed plausible so I went with DTE.

Now I did not think about DET(Data Export Tool) before. It can be confusing. Now I am also inclined to go with UCRs instead of DET, just want to confirm it would be a good idea.

[the remark this answers is not in the document]
Comment (snopoke) [reply corehq/apps/domain/models.py:1064]: I think sticking with UCR for now makes sense. Changing all instances to DTE is a big job and not clear what the value is.

File: corehq/apps/domain/utils.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (AmitPhulera) [reply corehq/apps/domain/utils.py:129]: It contains the list of allowed UCR expressions for a domain.

> What is the logic to get the list of restricted expressions for a domain?

First thing is that the change only affects two expressions `BaseItemExpression` and `RelatedDocumentExpressions`.  If a user tries to create any of the above expressions, we will see if the domain is allowed to use the expression. For that, we will query `AllowedDTEExpressionSettings`. If the expression is present in the model then we will allow it else we will raise an error.

[the remark this answers is not in the document]
Comment (snopoke) [reply corehq/apps/domain/utils.py:129]: OK, I think it's making sense now. I think it would actually be better to store the list of disallowed expressions in the DB. The reason I think that is because it makes it easier to add more expressions to the list in future without having to do a migration.

[the remark this answers is not in the document]
Comment (AmitPhulera) [reply corehq/apps/domain/utils.py:129]: The reason we are saving only allowed attributes is because -
- It translates well with the UI as we have to manually add the allowed expressions to the list.
- The list is unlikely to grow in the future, the expressions that are being restricted seemed to have a platform. Where abusing the expression could have caused performance issues. We can work on making them GA in the future but things going other way round is highly unlikely.

File: corehq/apps/domain/models.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (AmitPhulera) [reply corehq/apps/domain/models.py:1064]: I have renamed DTE to UCR.

File: corehq/apps/domain/utils.py [hunk 3]
```diff
@@ -124,3 +124,22 @@ def get_serializable_wire_invoice_general_credit(general_credit):
         }]
 
     return []
+
+
+@quickcache(['domain_name'])
+def get_allowed_ucr_expressions(domain_name):
```
Comment (biyeun) [inline corehq/apps/domain/utils.py:130]: does it make sense to add `get_allowed_ucr_expressions` and `save_allowed_ucr_expressions` as `classmethod`s to `AllowedUCRExpressionSettings` since it's very specific to that model?

File: corehq/apps/domain/utils.py [hunk 4]
```diff
@@ -124,3 +124,22 @@ def get_serializable_wire_invoice_general_credit(general_credit):
...
+    try:
+        ucr_expressions_obj = AllowedUCRExpressionSettings.objects.get(domain=domain_name)
+        allowed_ucr_expressions = ucr_expressions_obj.allowed_ucr_expressions
+    except AllowedUCRExpressionSettings.DoesNotExist:
+        allowed_ucr_expressions = get_default_ucr_expressions()
+    return allowed_ucr_expressions
+
+
+def save_allowed_ucr_expressions(domain_name, expressions):
```
Comment (biyeun) [inline corehq/apps/domain/utils.py:139]: this very much seems like it should be a `classmethod` on `AllowedUCRExpressionSettings`. also, is it worth adding a test? Are we planning to audit any of these changes?

Comment (AmitPhulera) [reply corehq/apps/domain/utils.py:139]: I have added tests for `get_allowed_ucr_expressions` https://github.com/dimagi/commcare-hq/blob/1992eed95179e6b989418415d6885a159e9e8d21/corehq/apps/domain/tests/test_utils.py#L78-L85

The`save_allowed_ucr_expressions ` seemed like a simple Model query so I don't think it requires any tests.

I was also thinking of moving them to the model class first. The reason I kept it in utils was that both of the functions were only used in forms and views, so decided to go with the utility function.

I would be happy to move it to the model class if that suits it better.

Comment (AmitPhulera) [reply corehq/apps/domain/utils.py]: > Are we planning to audit any of these changes?

I missed this earlier. I am not sure about it but it would be good if we do it. 
Should I add it?

File: corehq/apps/domain/models.py [hunk 5]
```diff
@@ -1047,6 +1059,46 @@ def update_calculations(cls, domain, property_to_update):
         obj.save(update_fields=[property_to_update])
 
 
+class AllowedUCRExpressionSettings(models.Model):
+    """
+    Model contains DTE(aka UCR) expressions settings for a domain.
+    The expressions defined in RESTRICTED_DTE_EXPRESSIONS are not generally available yet.
```
Comment (calellowitz) [inline corehq/apps/domain/models.py:1065]: Since we renamed all the variables to UCR, would be good to update here.

File: corehq/apps/domain/tests/test_utils.py [hunk 6]
```diff
@@ -70,3 +72,13 @@ def test_domain(name="domain", skip_full_delete=False):
             Domain.get_db().delete_doc(domain.get_id)
         else:
             domain.delete()
+
+
+class TestDTEExpressionUtils(TestCase):
```
Comment (calellowitz) [inline corehq/apps/domain/tests/test_utils.py:77]: nit: this should be UCR as well

File: corehq/apps/userreports/static/userreports/js/configurable_reports_home.js [hunk 7]
```diff
@@ -15,7 +15,7 @@ hqDefine("userreports/js/configurable_reports_home", [
     });
 
     $select.select2({
-        placeholder: gettext("Edit a report or data source"),
+        placeholder: gettext("Edit a custom web report or custom web report source"),
```
Comment (calellowitz) [inline corehq/apps/userreports/static/userreports/js/configurable_reports_home.js:18]: I think this text change adds confusion rather than remove it since the reports and data sources edited on this page can be used in web or mobile.

File: corehq/apps/userreports/templates/userreports/configurable_reports_home.html [hunk 8]
```diff
@@ -5,18 +5,18 @@
 {% requirejs_main 'userreports/js/configurable_reports_home' %}
 
 {% block page_content %}
-  <h1>{% trans "Configurable Reports" %}</h1>
+  <h1>{% trans "Custom Web Reports" %}</h1>
```
Comment (calellowitz) [inline corehq/apps/userreports/templates/userreports/configurable_reports_home.html:8]: Same comment about confusion. these are not web only reports, so adding this word makes it less accurate.

File: corehq/apps/userreports/templates/userreports/edit_data_source.html [hunk 9]
```diff
@@ -7,8 +7,8 @@
   {% if data_source.get_id %}
     <div class="btn-toolbar pull-right">
         <div class="btn-group">
-          <a href="{% url 'preview_configurable_data_source' domain data_source.get_id %}" class="btn btn-default">{% trans 'Preview Data' %}</a>
-          <a href="{% url 'summary_configurable_data_source' domain data_source.get_id %}" class="btn btn-default">{% trans 'Data Source Summary' %}</a>
+          <a href="{% url 'preview_configurable_data_source' domain data_source.get_id %}" class="btn btn-default">{% trans 'Preview Custom Web Report Data' %}</a>
```
Comment (calellowitz) [inline corehq/apps/userreports/templates/userreports/edit_data_source.html:10]: same language comment

File: corehq/apps/users/static/users/js/roles.js [hunk 10]
```diff
@@ -359,10 +359,10 @@ hqDefine('users/js/roles',[
                     }];
                 self.ucrs = [{
                     visibilityRestraint: self.permissions.access_all_locations,
-                    text: gettext("Create and Edit Configurable Reports"),
+                    text: gettext("Create and Edit Custom Web Reports"),
```
Comment (calellowitz) [inline corehq/apps/users/static/users/js/roles.js:362]: here too

File: corehq/apps/domain/models.py [hunk 11]
```diff
@@ -1047,6 +1059,46 @@ def update_calculations(cls, domain, property_to_update):
...
+    Domain settings page on HQ.
+    """
+
+    domain = models.CharField(unique=True, max_length=256)
+    allowed_ucr_expressions = ArrayField(
+        models.CharField(max_length=32, choices=RESTRICTED_UCR_EXPRESSIONS),
+        default=get_default_ucr_expressions
+    )
+
+    def save(self, *args, **kwargs):
```
Comment (calellowitz) [inline corehq/apps/domain/models.py:1078]: Why is this necessary? I would have thought a standard array field would allow us to use the standard save function.

**Review by calellowitz — COMMENTED** [review]
Left a few small comments. I am still a little concerned about the workflow for SaaS here. This seems really easy to forget or to mess up. It might be nice to add something that will safely enable UCRs for saas customers in a single place, so it can't be messed up, perhaps a button or form somewhere that will set up this model at the same time it enables the feature flag, leaving CS only one step they need to take, and making it harder to forget or do incorrectly.

**Comment by AmitPhulera** [comment]
Hey @calellowitz 
Thanks for the inputs, I will update the comment and the tests. 
But the rename on the UI are done as per the [spec](https://docs.google.com/document/d/19oscr4oHnaZiRE7IG6oyBsAR0kKLQViF4Y-ufXggjAg/edit) provided by the product.

File: corehq/apps/domain/models.py (hunk 11 above)
Comment (AmitPhulera) [reply corehq/apps/domain/models.py:1078]: This was required to clear cache of `get_allowed_ucr_expressions`

Comment (snopoke) [reply corehq/apps/domain/models.py:1078]: The standard way to do this would be to use a `post_save` signal handler

Comment (AmitPhulera) [reply corehq/apps/domain/models.py:1078]: Ohh yes!! I will update it.

Comment (AmitPhulera) [reply corehq/apps/domain/models.py]: https://github.com/dimagi/commcare-hq/pull/31103/commits/05855b3b38f76c40a25775120abfa1c482dea67e

**Comment by AmitPhulera** [comment]
> This seems really easy to forget or to mess up. It might be nice to add something that will safely enable UCRs for saas customers in a single place, so it can't be messed up, perhaps a button or form somewhere that will set up this model at the same time it enables the feature flag, leaving CS only one step they need to take, and making it harder to forget or do incorrectly.

Having a form that takes in the domain name and enables the feature flag for the domain and simultaneously add the permissions seems like a good idea. We can discuss it with the Product to see where it should fit.

**Comment by calellowitz** [comment]
> Hey @calellowitz Thanks for the inputs, I will update the comment and the tests. But the rename on the UI are done as per the [spec](https://docs.google.com/document/d/19oscr4oHnaZiRE7IG6oyBsAR0kKLQViF4Y-ufXggjAg/edit) provided by the product.

I am happy to chat with product about this, but I think what I said before is correct. These changes add confusion because the report configurations and data sources that are referred to in the changed text can be used in both web reports and mobile ones, so changing the language to explicitly say "Web" makes it less precise rather than moreso, and could confuse anyone already using mobile UCRs

File: corehq/apps/domain/utils.py (hunk 4 above)
Comment (biyeun) [reply corehq/apps/domain/utils.py]: i think this might be a good question for the auditablity pod (@joxl) but I feel that might be a good idea?

Comment (AmitPhulera) [reply corehq/apps/domain/utils.py]: It would be good. I looked into logging the changes but the only place where I could feel it could be done was in `post`
See https://github.com/dimagi/commcare-hq/pull/31103/commits/90d07a1ce66074997aa1cef4452b485cd375844b  https://github.com/dimagi/commcare-hq/pull/31103/commits/e0789ad3f6c996cde91a5edb0763a4691006c716

**Review by biyeun — APPROVED** [review]
